# Transaction Datasets

In [ ]:
import pandas as pd
import numpy as np

pd.set_option('display.max_rows', 100)
pd.set_option('display.width', 120)

In [ ]:
train_tr = pd.read_csv('train_transaction.csv')
test_tr = pd.read_csv('test_transaction.csv')

print(f'train: {train_tr.shape}   test: {test_tr.shape}')

train: (590540, 394)   test: (506691, 393)


In [ ]:
train_tr.head()

,TransactionID,isFraud,TransactionDT,TransactionAmt,ProductCD,card1,card2,card3,card4,card5,...,V330,V331,V332,V333,V334,V335,V336,V337,V338,V339
0,2987000,0,86400,68.5,W,13926,NaN,150.0,discover,142.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2987001,0,86401,29.0,W,2755,404.0,150.0,mastercard,102.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2987002,0,86469,59.0,W,4663,490.0,150.0,visa,166.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2987003,0,86499,50.0,W,18132,567.0,150.0,mastercard,117.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2987004,0,86506,50.0,H,4497,514.0,150.0,mastercard,102.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


In [ ]:
test_tr.head()

,TransactionID,TransactionDT,TransactionAmt,ProductCD,card1,card2,card3,card4,card5,card6,...,V330,V331,V332,V333,V334,V335,V336,V337,V338,V339
0,3663549,18403224,31.95,W,10409,111.0,150.0,visa,226.0,debit,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,3663550,18403263,49.00,W,4272,111.0,150.0,visa,226.0,debit,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,3663551,18403310,171.00,W,4476,574.0,150.0,visa,226.0,debit,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,3663552,18403310,284.95,W,10989,360.0,150.0,visa,166.0,debit,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,3663553,18403317,67.95,W,18018,452.0,150.0,mastercard,117.0,debit,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [ ]:
# note, column does not exist in test dataset (flag created for sake of training)
print('Fraud count:')
print(train_tr['isFraud'].value_counts(), '\n')

print('\nFraud rate:')
print(train_tr['isFraud'].value_counts(normalize=True) * 100)

Fraud count:
isFraud
0    569877
1     20663
Name: count, dtype: int64 


Fraud rate:
isFraud
0    96.500999
1     3.499001
Name: proportion, dtype: float64


In [ ]:
print('Train dtype counts:')
print(train_tr.dtypes.value_counts(), '\n')

print('Test dtype counts:')
print(test_tr.dtypes.value_counts())

Train dtype counts:
float64    376
object      14
int64        4
Name: count, dtype: int64 

Test dtype counts:
float64    376
object      14
int64        3
Name: count, dtype: int64


In [ ]:
print('Top 30 Missing (train)')
miss_train = train_tr.isnull().mean().sort_values(ascending=False)
print((miss_train[miss_train > 0].head(30) * 100).round(2), '\n')

print('Top 30 Missing (test)')
miss_test = test_tr.isnull().mean().sort_values(ascending=False)
print((miss_test[miss_test > 0].head(30) * 100).round(2))

Top 30 Missing (train)
dist2    93.63
D7       93.41
D13      89.51
D14      89.47
D12      89.04
D6       87.61
D9       87.31
D8       87.31
V153     86.12
V149     86.12
V141     86.12
V146     86.12
V154     86.12
V162     86.12
V142     86.12
V158     86.12
V161     86.12
V157     86.12
V138     86.12
V139     86.12
V148     86.12
V140     86.12
V155     86.12
V156     86.12
V147     86.12
V163     86.12
V143     86.12
V145     86.12
V144     86.12
V165     86.12
dtype: float64 

Top 30 Missing (test)
dist2    92.81
D7       88.13
D12      86.33
D8       85.33
D9       85.33
V140     85.04
V155     85.04
V153     85.04
V154     85.04
V149     85.04
V141     85.04
V146     85.04
V148     85.04
V162     85.04
V142     85.04
V158     85.04
V161     85.04
V157     85.04
V138     85.04
V139     85.04
V147     85.04
V156     85.04
V163     85.04
V165     84.99
V144     84.99
V143     84.99
V164     84.99
V145     84.99
V166     84.99
V151     84.99
dtype: float64


In [ ]:
print('avg missing by metadata block (train)')
for prefix in ['C', 'D', 'M', 'V']:
    cols = [c for c in train_tr.columns if c.startswith(prefix)]
    if cols:
        print(f'{prefix}*: {train_tr[cols].isnull().mean().mean() * 100:.2f}% ({len(cols)} total cols)')

print('\navg missing by metadata block (test)')
for prefix in ['C', 'D', 'M', 'V']:
    cols = [c for c in test_tr.columns if c.startswith(prefix)]
    if cols:
        print(f'{prefix}*: {test_tr[cols].isnull().mean().mean() * 100:.2f}% ({len(cols)} total cols)')

avg missing by metadata block (train)
C*: 0.00% (14 total cols)
D*: 58.15% (15 total cols)
M*: 49.92% (9 total cols)
V*: 43.04% (339 total cols)

avg missing by metadata block (test)
C*: 0.07% (14 total cols)
D*: 50.68% (15 total cols)
M*: 42.57% (9 total cols)
V*: 38.62% (339 total cols)


In [ ]:
print(train_tr[['TransactionAmt', 'TransactionDT']].describe())

       TransactionAmt  TransactionDT
count   590540.000000   5.905400e+05
mean       135.027176   7.372311e+06
std        239.162522   4.617224e+06
min          0.251000   8.640000e+04
25%         43.321000   3.027058e+06
50%         68.769000   7.306528e+06
75%        125.000000   1.124662e+07
max      31937.391000   1.581113e+07


In [ ]:
print(train_tr.groupby('isFraud')['TransactionAmt'].describe())

            count        mean         std    min     25%   50%    75%        max
isFraud                                                                         
0        569877.0  134.511665  239.395078  0.251  43.970  68.5  120.0  31937.391
1         20663.0  149.244779  232.212163  0.292  35.044  75.0  161.0   5191.000


In [ ]:
days_train = (train_tr['TransactionDT'].max() - train_tr['TransactionDT'].min()) / (60 * 60 * 24)
days_test = (test_tr['TransactionDT'].max() - test_tr['TransactionDT'].min()) / (60 * 60 * 24)

print(f'\ntrain spans ~{days_train:.1f} days, test spans ~{days_test:.1f} days')
print(f'train DT range: {train_tr['TransactionDT'].min()} - {train_tr['TransactionDT'].max()}')
print(f'test  DT range: {test_tr['TransactionDT'].min()} - {test_tr['TransactionDT'].max()}')


train spans ~182.0 days, test spans ~183.0 days
train DT range: 86400 - 15811131
test  DT range: 18403224 - 34214345


In [ ]:
cat_cols = (['ProductCD', 'card1', 'card2', 'card3', 'card4', 'card5', 'card6',
             'addr1', 'addr2', 'P_emaildomain', 'R_emaildomain'] +
            [f'M{i}' for i in range(1, 10)])

print('unique vals per col (train)')
for col in cat_cols:
    print(f'{col}: {train_tr[col].nunique(dropna=True)}')

unique vals per col (train)
ProductCD: 5
card1: 13553
card2: 500
card3: 114
card4: 4
card5: 119
card6: 4
addr1: 332
addr2: 74
P_emaildomain: 59
R_emaildomain: 60
M1: 2
M2: 2
M3: 2
M4: 3
M5: 2
M6: 2
M7: 2
M8: 2
M9: 2


In [ ]:
cat_cols = (['ProductCD', 'card1', 'card2', 'card3', 'card4', 'card5', 'card6',
             'addr1', 'addr2', 'P_emaildomain', 'R_emaildomain'] +
            [f'M{i}' for i in range(1, 10)])

print('unique vals per col (test)')
for col in cat_cols:
    print(f'{col}: {test_tr[col].nunique(dropna=True)}')

unique vals per col (test)
ProductCD: 5
card1: 13244
card2: 497
card3: 115
card4: 4
card5: 102
card6: 3
addr1: 292
addr2: 72
P_emaildomain: 60
R_emaildomain: 60
M1: 2
M2: 2
M3: 2
M4: 3
M5: 2
M6: 2
M7: 2
M8: 2
M9: 2


In [ ]:
print('ProductCD fraud rate')
print(train_tr.groupby('ProductCD')['isFraud'].agg(['count', 'mean']))

ProductCD fraud rate
            count      mean
ProductCD                  
C           68519  0.116873
H           33024  0.047662
R           37699  0.037826
S           11628  0.058996
W          439670  0.020399


In [ ]:
print('card1 fraud rate')
print(train_tr.groupby('card1')['isFraud'].agg(['count', 'mean']))

card1 fraud rate
       count  mean
card1             
1000       1   0.0
1001       3   0.0
1004       5   0.0
1005       1   0.0
1006       3   0.0
...      ...   ...
18390     64   0.0
18391      1   0.0
18393      3   0.0
18395     11   0.0
18396      1   0.0

[13553 rows x 2 columns]


In [ ]:
print('card2 fraud rate')
print(train_tr.groupby('card2')['isFraud'].agg(['count', 'mean']))

card2 fraud rate
       count      mean
card2                 
100.0   7570  0.012814
101.0    156  0.000000
102.0    332  0.000000
103.0   3468  0.099481
104.0    754  0.248011
...      ...       ...
596.0    209  0.000000
597.0     86  0.000000
598.0     70  0.000000
599.0    403  0.007444
600.0    194  0.015464

[500 rows x 2 columns]


In [ ]:
print('card3 fraud rate')
print(train_tr.groupby('card3')['isFraud'].agg(['count', 'mean']))

card3 fraud rate
       count      mean
card3                 
100.0    202  0.049505
101.0      1  0.000000
102.0    460  0.023913
105.0      3  0.000000
106.0   1571  0.015277
...      ...       ...
226.0      2  0.000000
227.0     17  0.000000
228.0      1  0.000000
229.0    118  0.008475
231.0      6  0.166667

[114 rows x 2 columns]


In [ ]:
print('card4 fraud rate')
print(train_tr.groupby('card4')['isFraud'].agg(['count', 'mean']))

card4 fraud rate
                   count      mean
card4                             
american express    8328  0.028698
discover            6651  0.077282
mastercard        189217  0.034331
visa              384767  0.034756


In [ ]:
print('card5 fraud rate')
print(train_tr.groupby('card5')['isFraud'].agg(['count', 'mean']))

card5 fraud rate
       count      mean
card5                 
100.0    978  0.018405
101.0      7  0.000000
102.0  29105  0.062945
104.0     23  0.000000
105.0      2  0.000000
...      ...       ...
233.0     19  0.000000
234.0      1  0.000000
235.0      6  0.000000
236.0    647  0.030912
237.0     49  0.040816

[119 rows x 2 columns]


In [ ]:
print('card6 fraud rate')
print(train_tr.groupby('card6')['isFraud'].agg(['count', 'mean']))

card6 fraud rate
                  count      mean
card6                            
charge card          15  0.000000
credit           148986  0.066785
debit            439938  0.024263
debit or credit      30  0.000000


In [ ]:
print('top 10 P_emaildomain fraud rate')

top_emails = train_tr['P_emaildomain'].value_counts().head(10).index

print(train_tr[train_tr['P_emaildomain'].isin(top_emails)]
                                  .groupby('P_emaildomain')['isFraud']
                                  .agg(['count', 'mean'])
                                  .sort_values('count', ascending=False))

top 10 P_emaildomain fraud rate
                count      mean
P_emaildomain                  
gmail.com      228355  0.043542
yahoo.com      100934  0.022757
hotmail.com     45250  0.052950
anonymous.com   36998  0.023217
aol.com         28289  0.021811
comcast.net      7888  0.031187
icloud.com       6267  0.031434
outlook.com      5096  0.094584
msn.com          4092  0.021994
att.net          4033  0.007439


In [ ]:
v_cols = [c for c in train_tr.columns if c.startswith('V')]

corr = train_tr[v_cols].corr().abs()
upper = corr.where(np.triu(np.ones(corr.shape), k=1).astype(bool))
highly_corr = [col for col in upper.columns if any(upper[col] > 0.95)]
perfect_corr = [col for col in upper.columns if any(upper[col] == 1)]

print(f'{len(v_cols)} total V columns in train')
print(f'{len(highly_corr)} are >0.95 correlated with another V column')
print(f'{len(perfect_corr)} are perfectly correlated with another V column')

339 total V columns in train
124 are >0.95 correlated with another V column
0 are perfectly correlated with another V column


In [ ]:
v_cols = [c for c in test_tr.columns if c.startswith('V')]

corr = test_tr[v_cols].corr().abs()
upper = corr.where(np.triu(np.ones(corr.shape), k=1).astype(bool))
highly_corr = [col for col in upper.columns if any(upper[col] > 0.95)]
perfect_corr = [col for col in upper.columns if any(upper[col] == 1)]

print(f'{len(v_cols)} total V columns in test')
print(f'{len(highly_corr)} are >0.95 correlated with another V column')
print(f'{len(perfect_corr)} are perfectly correlated with another V column')

339 total V columns in test
84 are >0.95 correlated with another V column
1 are perfectly correlated with another V column


## Notes:

*   20,663 cases of fraud positively flagged in train dataset (~3.5% of total)
    *   Might make assessing fraud detection accuracy a bad way to go for assessing model performance
    *   Maybe ROC-AUC or PR-AUC instead so we can catch false positives?
    *   May be good to consider adding some class weights when training models
*   Time gap between train and test: train runs day 0 to ~day 182, test runs ~day 213 to ~day 396, so there's about a 30-day gap between where train ends and test starts
    *   May cause trouble if we want to combine and re-split it on our own
          * Also would need to figure out how to create "isFraud" flag for new rows from test set, currently doesn't exist in test data
*   A lot of V columns have same missing pct rate; likely means if one col has a value present, it will also be present in many other V cols (haven't tested to confirm this 100% yet)
    *   High correlation between V cols
         *   Train: 124 cols >95% correlated to other V cols, 0 perfectly correlated
         *   Test: 84 cols >95% correlated, 1 perfect
    *  Wouldn't be the best to impute vals for this since over 80% of the vals are null
         *   How do we want to treat those null values? most models don't handle nulls very well
         *   Gradient boosted model uses nulls as a meaningful indicator instead of as a missing val (i think), maybe we should look into that?
    *   Maybe figure out if we can drop some of these cols for training to reduce noise
*   Want to look deeper into what unique values in each col are missing in train vs. in test:
    *   'card5' (119 train vs 102 test)
    *   'card6' (4 unique in train vs 3 in test)
    *   If they are rare cases, maybe lump into "other" col for easier processing?

*   Some cols that do show a significant difference in the amount of fraud detected b/t unique values:
    *   ProductCD, C 11.7% vs W 2.0%, big spread
    *   card4 and card6
    *   card2, card3, card5 may be too noisy, group or bin these?
    *   P_emaildomain
    *   Didn't check all so potentially others

## To-dos for model training pre-processing:
* Logistic regression:
    *   No missing values unless we find a library that can handle it. maybe use "missing" flag somehow instead of NaN/null?
    *   Regularization between numerical values
         *   TransactionAmt contains vals in the thousands (max is over 31k), operates on a different scale than other cols that may have a smaller range. need to get all columns operating on the same scale to make them more comparable or need to figure out how to make sure model is processing them correctly
         *   Figure out what we want to do with V cols in general
* Random forest:
    *   No missing vals also, gradient boosted package may address needs
    *   I don't think we need to worry about scaling/regularization
    *   590k rows, may need to think more about training time/compute

# Identity Datasets

In [ ]:
train_id = pd.read_csv('train_identity.csv')
test_id = pd.read_csv('test_identity.csv')

print(f'train_identity: {train_id.shape}   test_identity: {test_id.shape}')

train_identity: (144233, 41)   test_identity: (141907, 41)


In [ ]:
train_id.head()

,TransactionID,id_01,id_02,id_03,id_04,id_05,id_06,id_07,id_08,id_09,...,id_31,id_32,id_33,id_34,id_35,id_36,id_37,id_38,DeviceType,DeviceInfo
0,2987004,0.0,70787.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,samsung browser 6.2,32.0,2220x1080,match_status:2,T,F,T,T,mobile,SAMSUNG SM-G892A Build/NRD90M
1,2987008,-5.0,98945.0,NaN,NaN,0.0,-5.0,NaN,NaN,NaN,...,mobile safari 11.0,32.0,1334x750,match_status:1,T,F,F,T,mobile,iOS Device
2,2987010,-5.0,191631.0,0.0,0.0,0.0,0.0,NaN,NaN,0.0,...,chrome 62.0,NaN,NaN,NaN,F,F,T,T,desktop,Windows
3,2987011,-5.0,221832.0,NaN,NaN,0.0,-6.0,NaN,NaN,NaN,...,chrome 62.0,NaN,NaN,NaN,F,F,T,T,desktop,NaN
4,2987016,0.0,7460.0,0.0,0.0,1.0,0.0,NaN,NaN,0.0,...,chrome 62.0,24.0,1280x800,match_status:2,T,F,T,T,desktop,MacOS


In [ ]:
test_id.head()

,TransactionID,id-01,id-02,id-03,id-04,id-05,id-06,id-07,id-08,id-09,...,id-31,id-32,id-33,id-34,id-35,id-36,id-37,id-38,DeviceType,DeviceInfo
0,3663586,-45.0,280290.0,NaN,NaN,0.0,0.0,NaN,NaN,NaN,...,chrome 67.0 for android,NaN,NaN,NaN,F,F,T,F,mobile,MYA-L13 Build/HUAWEIMYA-L13
1,3663588,0.0,3579.0,0.0,0.0,0.0,0.0,NaN,NaN,0.0,...,chrome 67.0 for android,24.0,1280x720,match_status:2,T,F,T,T,mobile,LGLS676 Build/MXB48T
2,3663597,-5.0,185210.0,NaN,NaN,1.0,0.0,NaN,NaN,NaN,...,ie 11.0 for tablet,NaN,NaN,NaN,F,T,T,F,desktop,Trident/7.0
3,3663601,-45.0,252944.0,0.0,0.0,0.0,0.0,NaN,NaN,0.0,...,chrome 67.0 for android,NaN,NaN,NaN,F,F,T,F,mobile,MYA-L13 Build/HUAWEIMYA-L13
4,3663602,-95.0,328680.0,NaN,NaN,7.0,-33.0,NaN,NaN,NaN,...,chrome 67.0 for android,NaN,NaN,NaN,F,F,T,F,mobile,SM-G9650 Build/R16NW


In [ ]:
# fix col names
test_id.columns = [c.replace('-', '_') for c in test_id.columns]
test_id.head()

,TransactionID,id_01,id_02,id_03,id_04,id_05,id_06,id_07,id_08,id_09,...,id_31,id_32,id_33,id_34,id_35,id_36,id_37,id_38,DeviceType,DeviceInfo
0,3663586,-45.0,280290.0,NaN,NaN,0.0,0.0,NaN,NaN,NaN,...,chrome 67.0 for android,NaN,NaN,NaN,F,F,T,F,mobile,MYA-L13 Build/HUAWEIMYA-L13
1,3663588,0.0,3579.0,0.0,0.0,0.0,0.0,NaN,NaN,0.0,...,chrome 67.0 for android,24.0,1280x720,match_status:2,T,F,T,T,mobile,LGLS676 Build/MXB48T
2,3663597,-5.0,185210.0,NaN,NaN,1.0,0.0,NaN,NaN,NaN,...,ie 11.0 for tablet,NaN,NaN,NaN,F,T,T,F,desktop,Trident/7.0
3,3663601,-45.0,252944.0,0.0,0.0,0.0,0.0,NaN,NaN,0.0,...,chrome 67.0 for android,NaN,NaN,NaN,F,F,T,F,mobile,MYA-L13 Build/HUAWEIMYA-L13
4,3663602,-95.0,328680.0,NaN,NaN,7.0,-33.0,NaN,NaN,NaN,...,chrome 67.0 for android,NaN,NaN,NaN,F,F,T,F,mobile,SM-G9650 Build/R16NW


In [ ]:
train_tr_merge = train_tr[['TransactionID', 'isFraud']]
match_rate = train_tr_merge['TransactionID'].isin(train_id['TransactionID']).mean()
print(f'% of train transactions that have an identity record: {match_rate * 100:.2f}%')

% of train transactions that have an identity record: 24.42%


In [ ]:
merged = train_tr_merge.merge(train_id[['TransactionID']], on='TransactionID', how='left', indicator=True)
print('fraud rate, has identity record vs. does not')
print(train_tr_merge.assign(has_identity=train_tr_merge['TransactionID'].isin(train_id['TransactionID']))
      .groupby('has_identity')['isFraud'].agg(['count', 'mean']))

fraud rate, has identity record vs. does not
               count      mean
has_identity                  
False         446307  0.020939
True          144233  0.078470


In [ ]:
print('train')
print(train_id.dtypes.value_counts())

print('\ntest')
print(test_id.dtypes.value_counts())

train
float64    23
object     17
int64       1
Name: count, dtype: int64

test
float64    23
object     17
int64       1
Name: count, dtype: int64


In [ ]:
print('sorted missing vals train')
miss_train = train_id.isnull().mean().sort_values(ascending=False)
print((miss_train[miss_train > 0] * 100).round(2))

print('\nsorted missing vals train')
miss_test = test_id.isnull().mean().sort_values(ascending=False)
print((miss_test[miss_test > 0] * 100).round(2))

sorted missing vals train
id_24         96.71
id_25         96.44
id_07         96.43
id_08         96.43
id_21         96.42
id_26         96.42
id_23         96.42
id_27         96.42
id_22         96.42
id_18         68.72
id_04         54.02
id_03         54.02
id_33         49.19
id_10         48.05
id_09         48.05
id_30         46.22
id_32         46.21
id_34         46.06
id_14         44.50
DeviceInfo    17.73
id_13         11.73
id_16         10.33
id_06          5.11
id_05          5.11
id_20          3.45
id_19          3.41
id_17          3.37
id_31          2.74
DeviceType     2.37
id_02          2.33
id_28          2.26
id_29          2.26
id_11          2.26
id_38          2.25
id_35          2.25
id_36          2.25
id_15          2.25
id_37          2.25
dtype: float64

sorted missing vals train
id_24         96.66
id_25         96.45
id_26         96.44
id_08         96.43
id_07         96.43
id_21         96.43
id_23         96.43
id_27         96.43
id_22       

In [ ]:
# based on metadata
cat_cols = ['DeviceType', 'DeviceInfo'] + [f'id_{i}' for i in range(12, 39)]
cat_cols = [c for c in cat_cols if c in train_id.columns]

print('train set unique vals per column')
for col in cat_cols:
    print(f'{col}: {train_id[col].nunique(dropna=True)}')

train set unique vals per column
DeviceType: 2
DeviceInfo: 1786
id_12: 2
id_13: 54
id_14: 25
id_15: 3
id_16: 2
id_17: 104
id_18: 18
id_19: 522
id_20: 394
id_21: 490
id_22: 25
id_23: 3
id_24: 12
id_25: 341
id_26: 95
id_27: 2
id_28: 2
id_29: 2
id_30: 75
id_31: 130
id_32: 4
id_33: 260
id_34: 4
id_35: 2
id_36: 2
id_37: 2
id_38: 2


In [ ]:
print('test set unique vals per column')
for col in cat_cols:
    print(f'{col}: {test_id[col].nunique(dropna=True)}')

test set unique vals per column
DeviceType: 2
DeviceInfo: 2226
id_12: 2
id_13: 27
id_14: 27
id_15: 3
id_16: 2
id_17: 103
id_18: 17
id_19: 502
id_20: 436
id_21: 443
id_22: 26
id_23: 3
id_24: 15
id_25: 309
id_26: 94
id_27: 2
id_28: 2
id_29: 2
id_30: 86
id_31: 135
id_32: 5
id_33: 390
id_34: 2
id_35: 2
id_36: 2
id_37: 2
id_38: 2


In [ ]:
num_cols = [c for c in train_id.columns if c.startswith('id_') and c not in cat_cols]
print('numeric ID cols describe (train)')
print(train_id[num_cols].describe().T)

numeric ID cols describe (train)
          count           mean            std    min      25%       50%       75%       max
id_01  144233.0     -10.170502      14.347949 -100.0    -10.0      -5.0      -5.0       0.0
id_02  140872.0  174716.584708  159651.816856    1.0  67992.0  125800.5  228749.0  999595.0
id_03   66324.0       0.060189       0.598231  -13.0      0.0       0.0       0.0      10.0
id_04   66324.0      -0.058938       0.701015  -28.0      0.0       0.0       0.0       0.0
id_05  136865.0       1.615585       5.249856  -72.0      0.0       0.0       1.0      52.0
id_06  136865.0      -6.698710      16.491104 -100.0     -6.0       0.0       0.0       0.0
id_07    5155.0      13.285354      11.384207  -46.0      5.0      14.0      22.0      61.0
id_08    5155.0     -38.600388      26.084899 -100.0    -48.0     -34.0     -23.0       0.0
id_09   74926.0       0.091023       0.983842  -36.0      0.0       0.0       0.0      25.0
id_10   74926.0      -0.301124       2.789446 -

In [ ]:
num_cols = [c for c in test_id.columns if c.startswith('id_') and c not in cat_cols]
print('numeric ID cols describe (test)')
print(test_id[num_cols].describe().T)

numeric ID cols describe (test)
          count           mean            std    min      25%       50%       75%       max
id_01  141907.0     -11.325734      14.508520 -100.0    -12.5      -5.0      -5.0       0.0
id_02  136976.0  192658.729909  182613.277215    2.0  63339.5  133189.5  265717.5  999869.0
id_03   66481.0       0.053008       0.684551  -12.0      0.0       0.0       0.0      11.0
id_04   66481.0      -0.087454       0.840351  -19.0      0.0       0.0       0.0       0.0
id_05  134750.0       1.246033       5.071394  -81.0      0.0       0.0       1.0      52.0
id_06  134750.0      -6.803829      15.921457 -100.0     -6.0       0.0       0.0       0.0
id_07    5059.0      12.493180      11.678206  -41.0      3.0      12.0      21.0      59.0
id_08    5059.0     -36.577782      25.544185 -100.0    -46.0     -33.0     -23.0       0.0
id_09   74338.0       0.076219       1.009687  -32.0      0.0       0.0       0.0      16.0
id_10   74338.0      -0.245877       2.120525 -1

In [ ]:
print('DeviceType value counts (train)')
print(train_id['DeviceType'].value_counts(dropna=False))

print('\nDeviceInfo top 20 value counts (train)')
print(train_id['DeviceInfo'].value_counts(dropna=False).head(20))

DeviceType value counts (train)
DeviceType
desktop    85165
mobile     55645
NaN         3423
Name: count, dtype: int64

DeviceInfo top 20 value counts (train)
DeviceInfo
Windows                        47722
NaN                            25567
iOS Device                     19782
MacOS                          12573
Trident/7.0                     7440
rv:11.0                         1901
rv:57.0                          962
SM-J700M Build/MMB29K            549
SM-G610M Build/MMB29K            461
SM-G531H Build/LMY48B            410
rv:59.0                          362
SM-G935F Build/NRD90M            334
SM-G955U Build/NRD90M            328
SM-G532M Build/MMB29T            316
ALE-L23 Build/HuaweiALE-L23      312
SM-G950U Build/NRD90M            290
SM-G930V Build/NRD90M            274
rv:58.0                          269
rv:52.0                          256
SAMSUNG                          235
Name: count, dtype: int64


In [ ]:
print('DeviceType value counts (test)')
print(test_id['DeviceType'].value_counts(dropna=False))

print('\nDeviceInfo top 20 value counts (test)')
print(test_id['DeviceInfo'].value_counts(dropna=False).head(20))

DeviceType value counts (test)
DeviceType
desktop    74403
mobile     62528
NaN         4976
Name: count, dtype: int64

DeviceInfo top 20 value counts (test)
DeviceInfo
Windows                        44988
NaN                            26850
iOS Device                     18720
MacOS                          11149
Trident/7.0                     4890
rv:11.0                          749
SM-G532M Build/MMB29T            664
rv:63.0                          474
rv:61.0                          419
SM-G610M Build/MMB29K            343
rv:62.0                          335
ANE-LX3                          315
rv:64.0                          310
moto g(6) play                   295
SM-J700M Build/MMB29K            280
Moto G (5) Plus                  272
SM-G531H Build/LMY48B            272
SM-G950U                         252
Moto G (5)                       244
ANE-LX3 Build/HUAWEIANE-LX3      239
Name: count, dtype: int64


In [ ]:
merged_full = train_id.merge(train_tr_merge, on='TransactionID', how='left')

print('fraud rate by DeviceType')
print(merged_full.groupby('DeviceType', dropna=False)['isFraud'].agg(['count', 'mean']))

print('\nfraud rate by top 10 DeviceInfo')
top_devices = merged_full['DeviceInfo'].value_counts().head(10).index
print(merged_full[merged_full['DeviceInfo'].isin(top_devices)]
      .groupby('DeviceInfo')['isFraud'].agg(['count', 'mean'])
      .sort_values('count', ascending=False))

fraud rate by DeviceType
            count      mean
DeviceType                 
desktop     85165  0.065215
mobile      55645  0.101662
NaN          3423  0.031259

fraud rate by top 10 DeviceInfo
                       count      mean
DeviceInfo                            
Windows                47722  0.065400
iOS Device             19782  0.062683
MacOS                  12573  0.022111
Trident/7.0             7440  0.012903
rv:11.0                 1901  0.039979
rv:57.0                  962  0.107069
SM-J700M Build/MMB29K    549  0.109290
SM-G610M Build/MMB29K    461  0.080260
SM-G531H Build/LMY48B    410  0.080488
rv:59.0                  362  0.121547


In [ ]:
print(f'in train not test: {set(train_id.columns) - set(test_id.columns)}')
print(f'in test not train: {set(test_id.columns) - set(train_id.columns)}')

for col in ['DeviceInfo', 'id_30', 'id_31', 'id_33']:
    if col in train_id.columns and col in test_id.columns:
        new_vals = set(test_id[col].dropna().unique()) - set(train_id[col].dropna().unique())
        total = test_id[col].nunique(dropna=True)
        pct = 100 * len(new_vals) / total if total else 0
        print(f'{col}: {len(new_vals)} values in test never seen in train ({pct:.1f}% of test\'s unique values)')

in train not test: set()
in test not train: set()
DeviceInfo: 1013 values in test never seen in train (45.5% of test's unique values)
id_30: 12 values in test never seen in train (14.0% of test's unique values)
id_31: 42 values in test never seen in train (31.1% of test's unique values)
id_33: 201 values in test never seen in train (51.5% of test's unique values)


## Notes: Identity Data

*   Only 24.4% of transactions have an identity record at all
    *   Fraud rate is way different depending on this: 7.85% (has record) vs 2.09% (no record)
    *   "has_identity" flag before merging maybe?
*   Different levels of missing-ness
    *   id_24, id_25, id_07, id_08, id_21, id_26, id_23, id_27, id_22: ~96% missing
    *   id_18, id_03/04, id_33/09/10/30/32/34, id_14: 44-68%, normal imputation maybe
    *   Everything else under ~12%
*   DeviceInfo and a few id_ cols have some differences in unique val counts between test and train sets
    *   DeviceInfo: 1,786 train vs 2,226 test unique vals
    *   id_20 (394/436), id_25 (341/309), id_33 (260/390), id_31 (130/135)
    *   not an exhaustive list
*   Unseen category problem is bad here
    *   DeviceInfo: 45.5% of test's unique vals never seen in train
    *   id_33: 51.5%
    *   id_31: 31.1%
    *   id_30: 14.0%
    *   explicit "unknown/other" fallback?
*   id_11 basically no variance (90-100, mean 99.75, std 1.13), check if it's worth keeping
*   Numeric id_ cols look like bounded/clipped scores maybe?
    *   id_01, id_06, id_08 range -100 to 0
    *   id_10 -100 to 100 but clustered at 0
    *   might make more sense to bin these?

*   Cols to watch for predictability:
    *   DeviceType: desktop 6.5%, mobile 10.2%, missing 3.1%
    *   DeviceInfo grouped by OS family:  windows 6.5%, iOS 6.3%, macOS 2.2%, old IE 1.3%, some Android builds 8-12%

## To-dos for model training pre-processing: Identity Data
* Both models:
    *   Merge on TransactionID
    *   Group rare DeviceInfo/id_ categories into "other" before encoding, unseen category issue needs to be taken care of
* Logistic regression:
    *   Fill null values
    *   Standardize number columns
* Random forest:
    *   Fill null values